library/x509write_csr.c

```diff
@@ -85,6 +86,100 @@ int mbedtls_x509write_csr_set_extension(mbedtls_x509write_csr *ctx,
                                       critical, val, val_len);
 }
 
+int mbedtls_x509write_csr_set_subject_alternative_name(mbedtls_x509write_csr *ctx,
+                                                       const mbedtls_x509_san_list *san_list)
+{
+    int ret = 0;
+    const mbedtls_x509_san_list *cur = san_list;
+    unsigned char *buf;
+    unsigned char *p;
+    size_t len;
+    size_t buflen = 0;
+
+    /* Determine the maximum size of the SubjectAltName list */
+    while (cur != NULL) {
+        /* Calculate size of the required buffer */
+        switch (cur->node.type) {
+            case MBEDTLS_X509_SAN_DNS_NAME:
+            case MBEDTLS_X509_SAN_UNIFORM_RESOURCE_IDENTIFIER:
+            case MBEDTLS_X509_SAN_IP_ADDRESS:
+                /* + length of value for each name entry,
+                 * + maximum 4 bytes for the length field,
+                 * + 1 byte for the tag/type.
+                 */
+                buflen += cur->node.san.unstructured_name.len + 4 + 1;
+                break;
+
+            default:
+                /* Not supported - skip. */
+                break;
+        }
+
+
+        cur = cur->next;
+    }
+
+    /* Add the extra length field and tag */
+    buflen += 4 + 1;
+
+    /* Allocate buffer */
+    buf = mbedtls_calloc(1, buflen);
+    if (buf == NULL) {
+        return MBEDTLS_ERR_ASN1_ALLOC_FAILED;
+    }
+
+    mbedtls_platform_zeroize(buf, buflen);
+    p = buf + buflen;
+
+    /* Write ASN.1-based structure */
+    cur = san_list;
+    len = 0;
+    while (cur != NULL) {
+        switch (cur->node.type) {
+            case MBEDTLS_X509_SAN_DNS_NAME:
+            case MBEDTLS_X509_SAN_UNIFORM_RESOURCE_IDENTIFIER:
+            case MBEDTLS_X509_SAN_IP_ADDRESS:
+                MBEDTLS_ASN1_CHK_CLEANUP_ADD(len,
+                                             mbedtls_asn1_write_raw_buffer(&p, buf,
+                                                                           (const unsigned char *)
```

Comment:
This line, split into three, is unreadable. I would suggest just moving the arguments back so that things fit well.

Reply:
I'm not sure if style checker will approve this.

Reply:
You're right, it doesn't :D let's ask an expert - @davidhorstmann-arm - is there a way to force the formatter to accept these lines not being split into several, even though they should not have any whitespaces inbetween?

Reply:
The formatter will try to split lines where it can so as to limit the width to 100 characters, but also aligning arguments.

If the lines as currently written are too long, you're allowed to place the arguments on the next line from the function name, one indent in, instead. I've checked that this passes code style locally:
```c
            case MBEDTLS_X509_SAN_IP_ADDRESS:
                MBEDTLS_ASN1_CHK_CLEANUP_ADD(
                    len,
                    mbedtls_asn1_write_raw_buffer(
                        &p, buf,
                        (const unsigned char *) cur->node.san.unstructured_name.p,
                        cur->node.san.unstructured_name.len));

```

Reply:
You can break lines after the opening parenthesis, then the arguments are only intended by 4 spaces rather than by the length of the function name plus opening parenthesis:
```
                MBEDTLS_ASN1_CHK_CLEANUP_ADD(
                    len,
                    mbedtls_asn1_write_raw_buffer(
                        &p, buf,
                        (const unsigned char *) cur->node.san.unstructured_name.p,
                        cur->node.san.unstructured_name.len));
                MBEDTLS_ASN1_CHK_CLEANUP_ADD(
                    len,
                    mbedtls_asn1_write_len(
                        &p, buf,
                        cur->node.san.unstructured_name.len));
                MBEDTLS_ASN1_CHK_CLEANUP_ADD(
                    len,
                    mbedtls_asn1_write_tag(
                        &p, buf,
                        MBEDTLS_ASN1_CONTEXT_SPECIFIC | cur->node.type));
```

Here, in addition, I think it would be good to give `cur->node.san.unstructured_name` a short name locally, and/or break this code into an auxiliary function.

Reply:
Thanks guys!